### PR TITLE
New option to consider acknowledgment at the entity level (#2772)

### PR DESCRIPTION
**Pull request by vlo-rte**
3 code smells but the first one is a little bit "strange", we can discuss it maybe. The 2 lasts are not on new code ....

Release notes : 

In Features section : 
#2772 : New option to consider acknowledgment at the entity level. Please see the [documentation about acknowledgment](https://opfab.github.io/documentation/current/reference_doc/#_card_acknowledgment) for more information.

### CODE REVIEW

File: ui/main/src/app/modules/share/light-card/light-card.component.ts [hunk 1]
```diff
@@ -35,15 +36,16 @@ export class LightCardComponent implements OnInit, OnDestroy {
...
     cardTitle: string;
     dateToDisplay: string;
     fromEntity = null;
-    showExpiredIcon: boolean = true;
-    showExpiredLabel: boolean = true;
-    expiredLabel: string = 'feed.lttdFinished';
+    showExpiredIcon = true;
+    showExpiredLabel = true;
+    expiredLabel = 'feed.lttdFinished';
+    currentUserWithPerimeters: UserWithPerimeters;
```
[the remark this answers is not in the document]
Comment (vlo-rte) [reply ui/main/src/app/modules/share/light-card/light-card.component.ts:48]: You're right. Done

File: ui/main/src/app/modules/share/light-card/light-card.component.spec.ts [hunk 2]
```diff
@@ -63,7 +66,8 @@ describe('LightCardComponent', () => {
                 {provide: Router, useValue: myrout},
                 ProcessesService,
                 {provide: 'TimeEventSource', useValue: null},
-                TimeService, I18nService
+                TimeService, I18nService,
+                {provide: CardService, useValue: cardServiceSpy},
```
[the remark this answers is not in the document]
Comment (vlo-rte) [reply ui/main/src/app/modules/share/light-card/light-card.component.spec.ts:70]: You're right. Done

File: ui/main/src/app/model/light-card.model.ts [hunk 3]
```diff
@@ -39,11 +39,14 @@ export class LightCard {
...
         readonly representativeType?: PublisherType | string,
+        readonly entitiesAcks?: string[],
+        readonly entityRecipients?: string[],
         readonly entitiesAllowedToRespond?: string[],
         readonly entitiesRequiredToRespond?: string[],
         readonly entitiesAllowedToEdit?: string[],
         readonly publisherType?: PublisherType | string,
-        readonly secondsBeforeTimeSpanForReminder?: number
+        readonly secondsBeforeTimeSpanForReminder?: number,
+        public acknowledgedByEntity: boolean = false
```
[the remark this answers is not in the document]
Comment (vlo-rte) [reply ui/main/src/app/model/light-card.model.ts:49]: Done

File: services/cards-publication/src/main/java/org/opfab/cards/publication/services/CardProcessingService.java [hunk 4]
```diff
@@ -355,7 +355,8 @@ public UserBasedOperationResult processUserAcknowledgement(String cardUid, User
                     .message("Acknowledgement impossible : User is not member of all the entities given in the request")
                     .build());
 
-        cardNotificationService.pushAckOfCardInRabbit(cardUid, entitiesAcks);
+        cardRepositoryService.findByUid(cardUid).ifPresent(selectedCard ->
+            cardNotificationService.pushAckOfCardInRabbit(cardUid, selectedCard.getId(), entitiesAcks));
```
[the remark this answers is not in the document]
Comment (vlo-rte) [reply services/cards-publication/src/main/java/org/opfab/cards/publication/services/CardProcessingService.java:360]: I think we already receive 404 http code when we try to ack a card with a uid that does not exist. See unit test : 
```
@Test
	void processUserAcknowledgementOfUnexistingCard() throws Exception {
		String cardUid = "NotExistingCardUid";		
		Optional <CardPublicationData> card = cardRepository.findByUid(cardUid);
		Assertions.assertThat(card).isNotPresent();
		mockMvc.perform(post("/cards/userAcknowledgement/" + cardUid)
				        .contentType(MediaType.APPLICATION_JSON).content("[]")).andExpect(status().isNotFound());
	}
```

File: ui/main/src/app/services/card.service.ts [hunk 5]
```diff
@@ -58,16 +58,16 @@ export class CardService {
 
     private selectedCardId: string = null;
 
-    private receivedAcksSubject = new Subject<{cardUid: string, entitiesAcks: string[]}>();
-    private receivedDisconnectedSubject = new Subject<boolean>()
+    private receivedAcksSubject = new Subject<{cardUid: string, cardId: string, entitiesAcks: string[]}>();
```
Comment (freddidierRTE) [inline ui/main/src/app/services/card.service.ts:61]: you do not need the cardId , i do not see any use in the code

Comment (vlo-rte) [reply ui/main/src/app/services/card.service.ts:61]: Done.

**Pull request merged**